src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -4,260 +4,112 @@ open Equinox.Core
         return toSlice page isLast }
 
     let private readLastEventAsync (reader: MessageDbReader) (streamName: string) (requiresLeader: bool) (eventType: string option) ct = task {
         let! events = reader.ReadLastEvent(streamName, requiresLeader, ct, ?eventType = eventType)
-        return toSlice events false }
+        return toSlice events true }
 
     let inline len (bytes: EventBody) = bytes.Length
     let private resolvedEventLen (x: ITimelineEvent<EventBody>) = len x.Data + len x.Meta
     let private resolvedEventBytes events = events |> Array.sumBy resolvedEventLen
-    let private loggedReadSlice reader streamName batchSize requiresLeader startPos batchIndex (log: ILogger) ct: Task<_> = task {
-        let act = Activity.Current
-        let! t, slice = readSliceAsync reader streamName batchSize startPos requiresLeader |> Stopwatch.time ct
-        let bytes, count = slice.Messages |> resolvedEventBytes, slice.Messages.Length
-        let reqMetric: Log.Measurement = { stream = streamName; interval = t; bytes = bytes; count = count}
-        let evt = Log.Slice reqMetric
-        if act <> null then act.IncMetric(count, bytes).AddLastVersion(slice.LastVersion) |> ignore
-        let log = if not (log.IsEnabled Events.LogEventLevel.Debug) then log else log |> Log.propResolvedEvents "Json" slice.Messages
-        (log |> Log.prop "startPos" startPos |> Log.prop "bytes" bytes |> Log.event evt).Information("Mdb{action:l} count={count} version={version}",
-            "Read", count, slice.LastVersion)
-        return slice }
-
-    let private readBatches (log: ILogger) fold originState tryDecode (readSlice: int64 -> int -> ILogger -> CancellationToken -> Task<StreamEventsSlice>)
-            (maxPermittedBatchReads: int option) (startPosition: int64) ct
-        : Task<int64 * 'state * int * int> =
-        let mutable batchCount, eventCount, pos = 0, 0, startPosition
+    let private readBatches fold originState tryDecode (readSlice: int64 -> CancellationToken -> Task<StreamEventsSlice>)
+            (maxPermittedBatchReads: int option) (fromVersion: int64) ct
+        : Task<int64 * 'state> =
+        let mutable batchCount, eventCount, pos = 0, 0, fromVersion
         let mutable version = -1L
         let mutable state = originState
         let rec loop () : Task<unit> = task {
-            match maxPermittedBatchReads with
-            | Some mpbr when batchCount >= mpbr -> log.Information "batch Limit exceeded"; invalidOp "batch Limit exceeded"
-            | _ -> ()
-
-            let batchLog = log |> Log.prop "batchIndex" batchCount
-            let! slice = readSlice pos batchCount batchLog ct
+            let! slice = readSlice pos ct
             version <- max version slice.LastVersion
             state <- slice.Messages |> Seq.chooseV tryDecode |> fold state
             batchCount <- batchCount + 1
             eventCount <- eventCount + slice.Messages.Length
             pos <- slice.LastVersion  + 1L
```

Comment:
need comment on this, or L80 or some header comment on version vs pos vs message_db version numbering

Reply:
it's inclusive. so if I load from v1 I'll get event 1,2,3. If I load from v2 I'll get 2,3 etc.

Reply:
(this was more aimed at L85)

if LastVersion is a StreamVersion (-1 based) and version is a version (0 based)
I've lost track of it all
For me
a) I'd try to keep as many layers 0 based as possible (and refer to those as `version`)
b) depending on how relevant the internal store versioning scheme is, I'd try to have a layer that deals with that (and use a separate name, e.g. `streamVersion`, but ideally something that sticks out and makes you think)

If a `pos` is a `version`, and reading at pos 1 is reading events to apply on top of the snapshot version 1 which is based on one event, it's beautiful for me